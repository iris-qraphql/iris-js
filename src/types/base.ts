export type Printable = {
  toString: () => string;
  toJSON: () => string;
  decode: <V>(v: unknown) => V;
};
