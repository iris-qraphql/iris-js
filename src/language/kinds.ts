/**
 * The set of allowed kind values for AST nodes.
 */
export enum Kind {
  /** Name */
  NAME = 'Name',

  /** Document */
  DOCUMENT = 'Document',
  ARGUMENT = 'Argument',

  /** Values */
  INT = 'IntValue',
  FLOAT = 'FloatValue',
  STRING = 'StringValue',
  BOOLEAN = 'BooleanValue',
  NULL = 'NullValue',
  ENUM = 'EnumValue',
  LIST = 'ListValue',
  OBJECT = 'ObjectValue',
  OBJECT_FIELD = 'ObjectField',

  /** Directives */
  DIRECTIVE = 'Directive',

  /** Types */
  NAMED_TYPE = 'NamedType',
  LIST_TYPE = 'ListType',
  NON_NULL_TYPE = 'NonNullType',

  ARGUMENT_DEFINITION = 'InputValueDefinition',
  /** Type Definitions */
  FIELD_DEFINITION = 'FieldDefinition',
  RESOLVER_TYPE_DEFINITION = 'ResolverTypeDefinition',
  DATA_TYPE_DEFINITION = 'DataTypeDefinition',
  VARIANT_DEFINITION = 'VariantDefinition',

  /** Directive Definitions */
  DIRECTIVE_DEFINITION = 'DirectiveDefinition',
}

/**
 * The Enum type representing the possible kind values of AST nodes.
 *
 * @deprecated Please use `Kind`. Will be remove in v17.
 */
export type KindEnum = typeof Kind;
