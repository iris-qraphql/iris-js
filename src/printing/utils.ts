import type { Maybe } from '../utils/type-level';

/**
 * Given maybeArray, print an empty string if it is null or empty, otherwise
 * print all items together separated by separator if provided
 */
export function join(
  maybeArray: Maybe<ReadonlyArray<string | undefined>>,
  separator = '',
): string {
  return maybeArray?.filter((x) => x).join(separator) ?? '';
}

/**
 * Given array, print each item on its own line, wrapped in an indented `{ }` block.
 */
export const block = (array: ReadonlyArray<string | undefined>) =>
  array.length > 0 ? `{\n${indent(join(array, '\n'))}\n}` : '{}';

/**
 * If maybeString is not null or empty, then wrap with start and end, otherwise print an empty string.
 */
export function wrap(
  start: string,
  maybeString: Maybe<string>,
  end: string = '',
): string {
  return maybeString != null && maybeString !== ''
    ? start + maybeString + end
    : '';
}

export function indent(str: string): string {
  return wrap('  ', str.replace(/\n/g, '\n  '));
}
