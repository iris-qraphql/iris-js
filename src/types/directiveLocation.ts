export enum IrisDirectiveLocation {
  // LEGACY
  QUERY = 'QUERY',
  MUTATION = 'MUTATION',
  SUBSCRIPTION = 'SUBSCRIPTION',
  FIELD = 'FIELD',
  FRAGMENT_DEFINITION = 'FRAGMENT_DEFINITION',
  FRAGMENT_SPREAD = 'FRAGMENT_SPREAD',
  INLINE_FRAGMENT = 'INLINE_FRAGMENT',
  // IRIS
  DATA_DEFINITION = 'DATA_DEFINITION',
  VARIANT_DEFINITION = 'VARIANT_DEFINITION',
  FIELD_DEFINITION = 'FIELD_DEFINITION',
  ARGUMENT_DEFINITION = 'ARGUMENT_DEFINITION',
}
