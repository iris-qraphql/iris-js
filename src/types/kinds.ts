export enum IrisKind {
  ARGUMENT_DEFINITION = 'InputValueDefinition',
  NAMED_TYPE = 'NamedType',
  LIST_TYPE = 'ListType',
  MAYBE_TYPE = 'MaybeType',
  FIELD_DEFINITION = 'FieldDefinition',
  VARIANT_DEFINITION = 'VariantDefinition',
  TYPE_DEFINITION = 'TypeDefinition',
  DIRECTIVE_DEFINITION = 'DirectiveDefinition',
  DOCUMENT = 'Document',
}

export enum TokenKind {
  SOF = '<SOF>',
  EOF = '<EOF>',
  DOLLAR = '$',
  AMP = '&',
  PAREN_L = '(',
  PAREN_R = ')',
  SPREAD = '...',
  COLON = ':',
  EQUALS = '=',
  AT = '@',
  BRACKET_L = '[',
  BRACKET_R = ']',
  BRACE_L = '{',
  PIPE = '|',
  BRACE_R = '}',
  NAME = 'Name',
  INT = 'Int',
  FLOAT = 'Float',
  STRING = 'String',
  BLOCK_STRING = 'BlockString',
  COMMENT = 'Comment',
  QUESTION_MARK = '?',
  BANG = '!',
}

export const scalarNames = ['String', 'Int', 'Float', 'ID', 'Boolean'];
