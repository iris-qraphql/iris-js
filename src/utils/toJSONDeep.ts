import { isObjectLike } from '../jsutils/isObjectLike';
import { mapValue } from '../jsutils/mapValue';

export function toJSONDeep(value: unknown): unknown {
  if (!isObjectLike(value)) {
    return value;
  }

  if (typeof value.toJSON === 'function') {
    return value.toJSON();
  }

  if (Array.isArray(value)) {
    return value.map(toJSONDeep);
  }

  return mapValue(value, toJSONDeep);
}

export function toJSONError(fn: () => unknown) {
  try {
    fn();
    return undefined;
  } catch (error) {
    return toJSONDeep(error);
  }
}
