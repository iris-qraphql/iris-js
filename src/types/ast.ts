import type { Kind } from 'graphql';
import { Token } from 'graphql';

import type { Source } from '../parsing/source';
import type { Maybe } from '../utils/type-level';

import { GQLKind, IrisDirectiveLocation, IrisKind } from './kinds';

export { specifiedScalarTypes } from 'graphql';

export type ASTNode =
  | NameNode
  | DocumentNode
  | ArgumentNode
  | StringValueNode
  | NullValueNode
  | DirectiveNode
  | NamedTypeNode
  | ListTypeNode
  | FieldDefinitionNode
  | ArgumentDefinitionNode
  | TypeDefinitionNode
  | VariantDefinitionNode
  | MaybeTypeNode;

export type ValueNode = NullValueNode | StringValueNode;

/**
 * @internal
 */
export const QueryDocumentKeys: {
  [NodeT in ASTNode as NodeT['kind']]: ReadonlyArray<keyof NodeT>;
} = {
  Name: [],
  Document: ['definitions'],
  Argument: ['name', 'value'],
  StringValue: [],
  NullValue: [],
  Directive: ['name', 'arguments'],
  NamedType: ['name'],
  ListType: ['type'],
  MaybeType: ['type'],
  FieldDefinition: ['description', 'name', 'type', 'directives'],
  InputValueDefinition: [
    'description',
    'name',
    'type',
    'defaultValue',
    'directives',
  ],
  TypeDefinition: ['description', 'name', 'directives', 'variants'],
  VariantDefinition: ['description', 'name', 'fields', 'directives'],
};

const kindValues = new Set<string>(Object.keys(QueryDocumentKeys));
/**
 * @internal
 */
export function isNode(maybeNode: any): maybeNode is ASTNode {
  const maybeKind = maybeNode?.kind;
  return typeof maybeKind === 'string' && kindValues.has(maybeKind);
}

export class Location {
  /**
   * The character offset at which this Node begins.
   */
  readonly start: number;

  /**
   * The character offset at which this Node ends.
   */
  readonly end: number;

  /**
   * The Token at which this Node begins.
   */
  readonly startToken: Token;

  /**
   * The Token at which this Node ends.
   */
  readonly endToken: Token;

  /**
   * The Source document the AST represents.
   */
  readonly source: Source;

  constructor(startToken: Token, endToken: Token, source: Source) {
    this.start = startToken.start;
    this.end = endToken.end;
    this.startToken = startToken;
    this.endToken = endToken;
    this.source = source;
  }

  get [Symbol.toStringTag]() {
    return 'Location';
  }

  toJSON(): { start: number; end: number } {
    return { start: this.start, end: this.end };
  }
}

// VALUE
export type StringValueNode = {
  readonly kind: Kind.STRING;
  readonly loc?: Location;
  readonly value: string;
  readonly block?: boolean;
};

export type NullValueNode = {
  readonly kind: Kind.NULL;
  readonly loc?: Location;
};

export type ArgumentNode = {
  readonly kind: Kind.ARGUMENT;
  readonly loc?: Location;
  readonly name: NameNode;
  readonly value: ValueNode;
};

export type DirectiveNode = {
  readonly kind: Kind.DIRECTIVE;
  readonly loc?: Location;
  readonly name: NameNode;
  readonly arguments?: ReadonlyArray<ArgumentNode>;
};

/** Name */

export type NameNode = {
  readonly kind: Kind.NAME;
  readonly loc?: Location;
  readonly value: string;
};

/** Document */

export type DocumentNode = {
  readonly kind: IrisKind.DOCUMENT;
  readonly loc?: Location;
  readonly definitions: ReadonlyArray<TypeDefinitionNode>;
};

export { Token };

/** Type Reference */

export type WrapperKind = 'LIST' | 'MAYBE' | 'NAMED';

export type TypeNode = NamedTypeNode | ListTypeNode | MaybeTypeNode;

export type NamedTypeNode = {
  readonly kind: IrisKind.NAMED_TYPE;
  readonly loc?: Location;
  readonly name: NameNode;
};

export type ListTypeNode = {
  readonly kind: IrisKind.LIST_TYPE;
  readonly loc?: Location;
  readonly type: TypeNode;
};

export type MaybeTypeNode = {
  readonly kind: IrisKind.MAYBE_TYPE;
  readonly loc?: Location;
  readonly type: NamedTypeNode | ListTypeNode;
};

export type ArgumentDefinitionNode = {
  readonly kind: IrisKind.ARGUMENT_DEFINITION;
  readonly loc?: Location;
  readonly description?: StringValueNode;
  readonly name: NameNode;
  readonly type: TypeNode;
  readonly defaultValue?: ValueNode;
  readonly directives?: ReadonlyArray<DirectiveNode>;
};

export type FieldDefinitionNode = {
  readonly kind: IrisKind.FIELD_DEFINITION;
  readonly loc?: Location;
  readonly description?: StringValueNode;
  readonly name: NameNode;
  readonly type: TypeNode;
  readonly directives?: ReadonlyArray<DirectiveNode>;
};

export type VariantDefinitionNode = {
  readonly kind: IrisKind.VARIANT_DEFINITION;
  readonly loc?: Location;
  readonly description?: StringValueNode;
  readonly directives?: ReadonlyArray<DirectiveNode>;
  readonly isTypeVariantNode: boolean;
  readonly name: NameNode;
  readonly fields?: ReadonlyArray<FieldDefinitionNode>;
};

export type TypeDefinitionNode = {
  readonly kind: IrisKind.TYPE_DEFINITION;
  readonly loc?: Location;
  readonly description?: StringValueNode;
  readonly name: NameNode;
  readonly directives?: ReadonlyArray<DirectiveNode>;
  readonly variants: ReadonlyArray<VariantDefinitionNode>;
};

export interface DirectiveDefinitionNode {
  readonly kind: IrisKind.DIRECTIVE_DEFINITION;
  readonly loc?: Location;
  readonly description?: StringValueNode;
  readonly name: NameNode;
  readonly arguments?: ReadonlyArray<ArgumentDefinitionNode>;
  readonly repeatable: boolean;
  readonly locations: ReadonlyArray<NameNode>;
}

export const isTypeDefinitionNode = (
  node: ASTNode,
): node is TypeDefinitionNode => node.kind === IrisKind.TYPE_DEFINITION;

export const isTypeVariantNode = ({ name, variants }: TypeDefinitionNode) => {
  const typeName = name.value;
  const [variant] = variants;

  if (variants.length === 0) {
    return true;
  }

  return (
    variants.length === 1 &&
    variant.name.value === typeName &&
    variant.fields !== undefined
  );
};

export const getRefTypeName = (node: TypeNode): NameNode => {
  switch (node.kind) {
    case IrisKind.NAMED_TYPE:
      return node.name;
    case IrisKind.LIST_TYPE:
      return getRefTypeName(node.type);
    case IrisKind.MAYBE_TYPE:
      return getRefTypeName(node.type);
  }
};

export const getVariant = (
  { variants }: TypeDefinitionNode,
  name?: string,
): Maybe<VariantDefinitionNode> =>
  name ? variants.find((v) => v.name.value === name) : variants[0];

export const isRequiredArgument = (arg: ArgumentDefinitionNode) =>
  arg.type.kind !== IrisKind.MAYBE_TYPE && arg.defaultValue === undefined;

export const getDeprecationReason = (
  node: ASTNode & Pick<TypeDefinitionNode, 'directives'>,
): Maybe<string> => {
  const deprecated = node.directives?.find(
    (d) => d.name.value === 'deprecated',
  );

  if (!deprecated) {
    return undefined;
  }

  const reason = deprecated.arguments?.[0]?.value as Maybe<StringValueNode>;
  return reason?.value ?? '';
};

type IrisDirective = {
  name: string;
  description?: string;
  locations: ReadonlyArray<IrisDirectiveLocation>;
  args: ReadonlyArray<ArgumentDefinitionNode>;
  astNode?: DirectiveDefinitionNode;
};

export const specifiedDirectives: ReadonlyArray<IrisDirective> = Object.freeze([
  {
    name: 'deprecated',
    description: 'Marks an element of a GraphQL schema as no longer supported.',
    locations: [
      IrisDirectiveLocation.FIELD_DEFINITION,
      IrisDirectiveLocation.ARGUMENT_DEFINITION,
      IrisDirectiveLocation.VARIANT_DEFINITION,
    ],
    args: [
      {
        kind: IrisKind.ARGUMENT_DEFINITION,
        name: { value: 'reason', kind: GQLKind.NAME },
        description: {
          kind: GQLKind.STRING,
          value:
            'Explains why this element was deprecated, usually also including a suggestion for how to access supported similar data. Formatted using the Markdown syntax, as specified by [CommonMark](https://commonmark.org/).',
        },
        defaultValue: { value: '', kind: GQLKind.STRING },
        type: {
          kind: IrisKind.NAMED_TYPE,
          name: { value: 'string', kind: GQLKind.NAME },
        },
      },
    ],
  },
]);
