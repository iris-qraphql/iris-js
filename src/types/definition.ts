import type { GraphQLScalarType } from 'graphql';
import { specifiedScalarTypes } from 'graphql';
import { pluck } from 'ramda';

import { irisError } from '../error';
import { didYouMean, inspect, suggestionList } from '../utils/legacy';
import type { ObjMap } from '../utils/ObjMap';
import { keyMap, mapValue } from '../utils/ObjMap';
import type { Maybe } from '../utils/type-level';

import type {
  ArgumentDefinitionNode,
  FieldDefinitionNode,
  Role,
  TypeDefinitionNode,
  VariantDefinitionNode,
  WrapperKind,
} from './ast';
import { isTypeVariantNode } from './ast';
import type { Printable } from './base';
import { IrisTypeRefImp } from './type-ref';

export const stdScalars = keyMap(specifiedScalarTypes, ({ name }) => name);

export const scalarNames = Object.keys(stdScalars);

export const isSpecifiedScalarType = (type: IrisTypeDefinition): boolean =>
  Boolean(stdScalars[type.name]);

export const irisTypeRef = <K extends WrapperKind, T extends Printable>(
  kind: K,
  ofType: T,
) => new IrisTypeRefImp(kind, ofType);

export type IrisTypeRef<R extends Role = Role> =
  | IrisTypeRefImp<'LIST', IrisTypeRef<R>>
  | IrisTypeRefImp<'MAYBE', IrisTypeRef<R>>
  | IrisTypeRefImp<'NAMED', IrisTypeDefinition<R>>;

export const liftType = <R extends Role>(t: IrisTypeDefinition<R>) =>
  irisTypeRef<'NAMED', IrisTypeDefinition<R>>('NAMED', t);

type IrisNode = {
  name: string;
  description?: Maybe<string>;
  deprecationReason?: Maybe<string>;
  toJSON?: () => unknown;
};

export type ThunkObjMap<T> = Thunk<ObjMap<T>>;
export type Thunk<T> = (() => T) | T;

const isThunk = <T>(thunk: Thunk<T>): thunk is () => T =>
  typeof thunk === 'function';

export const resolveThunk = <T>(thunk: Thunk<T>): T =>
  isThunk(thunk) ? thunk() : thunk;

export type IrisArgument = IrisNode & {
  type: IrisTypeRef<'data'>;
  defaultValue?: unknown;
  astNode?: ArgumentDefinitionNode;
};

export const isRequiredArgument = (arg: IrisArgument): boolean =>
  arg.type.kind !== 'MAYBE' && arg.defaultValue === undefined;

export type IrisField<R extends Role = Role> = IrisNode & {
  astNode?: FieldDefinitionNode<R>;
  type: R extends 'data' ? IrisTypeRef<'data'> : IrisTypeRef<R>;
  args?: R extends 'data' ? never : ReadonlyArray<IrisArgument>;
};

export type IrisVariant<R extends Role = Role> = IrisNode & {
  astNode?: VariantDefinitionNode<R>;
  toJSON?: () => string;
  fields?: ObjMap<IrisField<R>>;
  type?: IrisTypeDefinition<R>;
};

type IrisTypeConfig<R extends Role> = {
  role: R;
  name: string;
  description?: Maybe<string>;
  variants: Thunk<ReadonlyArray<IrisVariant<R>>>;
  astNode?: Maybe<TypeDefinitionNode<R>>;
  scalar?: R extends 'data' ? GraphQLScalarType<any, any> : undefined;
};

export class IrisTypeDefinition<R extends Role = Role> {
  name: string;
  description: Maybe<string>;
  astNode: Maybe<TypeDefinitionNode<R>>;
  role: R;
  isVariantType: boolean;

  #thunkVariants: () => ReadonlyArray<IrisVariant<R>>;
  #resolvedVariants?: ReadonlyArray<IrisVariant<R>>;
  #scalar?: GraphQLScalarType;

  constructor(config: Readonly<IrisTypeConfig<R>>) {
    this.name = config.name;
    this.description = config.description;
    this.astNode = config.astNode;
    this.role = config.role;
    this.#thunkVariants = () => resolveThunk(config.variants);
    this.#scalar = config.scalar;
    this.isVariantType = config.astNode
      ? isTypeVariantNode(config.astNode)
      : false;
  }

  get [Symbol.toStringTag]() {
    return 'IrisTypeDefinition';
  }

  get boxedScalar() {
    if (this.#scalar) {
      return this.#scalar;
    }

    const [variant] = this.variants();
    return stdScalars[variant?.name];
  }

  variants = (): ReadonlyArray<IrisVariant<R>> => {
    if (this.#resolvedVariants) {
      return this.#resolvedVariants;
    }
    this.#resolvedVariants = this.#thunkVariants();

    return this.#resolvedVariants;
  };

  variantBy = (name?: string): IrisVariant<R> =>
    lookupVariant(this.name, this.variants(), name);

  toString = (): string => this.name;

  toJSON = (): string => this.toString();
}

const lookupVariant = <V extends { name: string }>(
  typeName: string,
  variants: ReadonlyArray<V>,
  name?: string,
): V => {
  if (!name) {
    if (variants.length !== 1) {
      throw irisError(
        `Object ${inspect(
          name,
        )} must provide variant name for type "${typeName}"`,
      );
    }
    return variants[0];
  }

  const variant = variants.find((v) => v.name === name);

  if (!variant) {
    throw irisError(
      `Data "${typeName}" cannot represent value: ${inspect(name)}` +
        didYouMean(
          'the variant',
          suggestionList(name, pluck('name', variants)),
        ),
    );
  }

  return variant;
};

export const IrisScalars = mapValue(
  stdScalars,
  (scalar) =>
    new IrisTypeDefinition({
      role: 'data',
      name: scalar.name,
      description: scalar.description,
      variants: [{ name: scalar.name }],
      scalar,
    }),
);
