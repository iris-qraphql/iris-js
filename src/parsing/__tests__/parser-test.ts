import { GQLKind } from '../../types/kinds';
import { toJSONDeep, toJSONError } from '../../utils/toJSONDeep';

import { parse, parseType, parseValue } from '../index';

export function expectJSON(actual: unknown) {
  const actualJSON = toJSONDeep(actual);
  return expect(actualJSON);
}

function expectSyntaxError(text: string) {
  return expect(toJSONError(() => parse(text)));
}

describe('Parser', () => {
  it('parse provides useful errors', () => {
    expectSyntaxError('data').toEqual({
      locations: [{ column: 5, line: 1 }],
      message: 'Syntax Error: Expected Name, found <EOF>.',
    });

    expectSyntaxError('...').toEqual({
      message: 'Syntax Error: Unexpected "...".',
      locations: [{ line: 1, column: 1 }],
    });
  });

  it('does not allow "true", "false", or "null" as Enum value', () => {
    expectSyntaxError('data Test = VALID | true ').toMatchSnapshot();
    expectSyntaxError('data Test = VALID | false').toMatchSnapshot();
    expectSyntaxError('data Test = VALID | null').toMatchSnapshot();
  });

  describe('parseValue', () => {
    it('parses null value', () => {
      const result = parseValue('null');
      expectJSON(result).toEqual({
        kind: GQLKind.NULL,
        loc: { start: 0, end: 4 },
      });
    });

    it('parses block strings', () => {
      expectJSON(parseValue('"""long"""')).toMatchSnapshot();
    });

    it('correct message for incomplete variable', () => {
      expect(() => parseValue('$')).toThrow();
      // .to.deep.include({
      //   message: 'Syntax Error: Expected Name, found <EOF>.',
      //   locations: [{ line: 1, column: 2 }],
      // });
    });

    it('correct message for unexpected token', () => {
      expect(() => parseValue(':')).toThrow();
      // .to.deep.include({
      //   message: 'Syntax Error: Unexpected ":".',
      //   locations: [{ line: 1, column: 1 }],
      // });
    });
  });

  describe('parseConstValue', () => {
    it('correct message for unexpected token', () => {
      expect(toJSONError(() => parseValue('$'))).toEqual({
        message: 'Syntax Error: Unexpected "$".',
        locations: [{ line: 1, column: 1 }],
      });
    });
  });

  describe('parseType', () => {
    it('parses well known types', () => {
      const result = parseType('String');
      expectJSON(result).toEqual({
        kind: GQLKind.NAMED_TYPE,
        loc: { start: 0, end: 6 },
        name: {
          kind: GQLKind.NAME,
          loc: { start: 0, end: 6 },
          value: 'String',
        },
      });
    });

    it('parses custom types', () => {
      const result = parseType('MyType');
      expectJSON(result).toEqual({
        kind: GQLKind.NAMED_TYPE,
        loc: { start: 0, end: 6 },
        name: {
          kind: GQLKind.NAME,
          loc: { start: 0, end: 6 },
          value: 'MyType',
        },
      });
    });

    it('parses list types', () => {
      const result = parseType('[MyType]');
      expectJSON(result).toEqual({
        kind: GQLKind.LIST_TYPE,
        loc: { start: 0, end: 8 },
        type: {
          kind: GQLKind.NAMED_TYPE,
          loc: { start: 1, end: 7 },
          name: {
            kind: GQLKind.NAME,
            loc: { start: 1, end: 7 },
            value: 'MyType',
          },
        },
      });
    });

    it('parses optional types', () => {
      const result = parseType('MyType?');
      expectJSON(result).toMatchSnapshot();
    });

    it('parses nested types', () => {
      const result = parseType('[MyType?]');
      expectJSON(result).toMatchSnapshot();
    });
  });
});
