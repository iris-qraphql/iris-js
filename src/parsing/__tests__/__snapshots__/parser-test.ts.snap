// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parser parseType parses nested types 1`] = `
Object {
  "kind": "ListType",
  "loc": Object {
    "end": 9,
    "start": 0,
  },
  "type": Object {
    "kind": "MaybeType",
    "loc": Object {
      "end": 8,
      "start": 1,
    },
    "type": Object {
      "kind": "NamedType",
      "loc": Object {
        "end": 7,
        "start": 1,
      },
      "name": Object {
        "kind": "Name",
        "loc": Object {
          "end": 7,
          "start": 1,
        },
        "value": "MyType",
      },
    },
  },
}
`;

exports[`Parser parseType parses optional types 1`] = `
Object {
  "kind": "MaybeType",
  "loc": Object {
    "end": 7,
    "start": 0,
  },
  "type": Object {
    "kind": "NamedType",
    "loc": Object {
      "end": 6,
      "start": 0,
    },
    "name": Object {
      "kind": "Name",
      "loc": Object {
        "end": 6,
        "start": 0,
      },
      "value": "MyType",
    },
  },
}
`;

exports[`Parser parseValue parses block strings 1`] = `
Object {
  "block": true,
  "kind": "StringValue",
  "loc": Object {
    "end": 10,
    "start": 0,
  },
  "value": "long",
}
`;
